arch/riscv/src/syscall.rs

```diff
@@ -328,7 +329,211 @@ impl kernel::syscall::UserspaceKernelBoundary for SysCall {
           : "memory"
           : "volatile");
 
-        let ret = match mcause::Trap::from(state.mcause as u32) {
+        #[cfg(feature = "riscv64")]
+        llvm_asm! ("
```

Comment:
This is probably the most concerning part of this PR for me. Here, the _middle_ of a function is being replaced with build system features. So even if a user checked the dependency tree, the function calls, and the function signatures, they _still_ wouldn't see any differences between the 32 and 64 bit versions. Using features this way means that any line of code can be different, and that is a rather large burden for future readers of the code. And, future readers would just have to have a hunch that two different platforms are compiling different code, because they otherwise are using the same libraries and same function calls.

My sense is the riscv designers intentionally didn't include a "load/store register" (pseudo)instruction. To me, that suggests that we should treat rv32i and rv64i separately. And rather than hiding that separation in the build system, we should use the structure that cargo/rust gives us and expose rv32i and rv64i crates. We can have a common arch/riscv crate for any shared code, but arch-specific code should live in its own crate (which also provides a place for specific documentation). It also future proofs things if we run into future differences between the architectures (or their physical implementations).

Reply:
C code commonly just sets a macro for the `sd`/`sw` commands. Would that be something we could do in Rust?

Having a separate `rv32i` and `rv64i` crate will result in lots of duplicated code, which risks having different bugs depending on XLEN. It would be best to try and avoid that.

32-bit and 64-bit RISC-V are very similar, so splitting them out isn't a great long term solution. As we add more RISC-V features the amount of duplicated code will continue to grow as well.